appscale/tools/node_layout.py

```diff
@@ -526,59 +526,52 @@ def to_list(self):
 
   def from_locations_json_list(self, locations_nodes_list):
     """Returns a list of nodes if the previous locations JSON matches with the
-    current NodeLayout from the AppScalefile. Otherwise returns None."""
+    current NodeLayout from the AppScalefile.
+
+    Args:
+      locations_nodes_list: A list of nodes in dictionary form loaded from
+        the locations json.
+    Raises:
+      AssertionError if the locations json nodes cannot be matched with the
+        AppScalefile nodes.
+    """
 
     # If the length does not match up the user has added or removed a node in
     # the AppScalefile.
-    if len(locations_nodes_list) != len(self.nodes):
-      return None
-
-    nodes = []
-
-    # Use a copy so we do not overwrite self.nodes when we call
-    # Node.from_json since that method modifies the node it is called on.
-    nodes_copy = self.nodes[:]
-    open_nodes = []
-    for old_node in locations_nodes_list:
-      old_node_roles = old_node.get('jobs')
-      # Convert deprecated roles to new roles.
-      old_node_roles = [role if role not in self.DEPRECATED_ROLES else
-                        self.DEPRECATED_ROLES[role] for role in old_node_roles]
-      if old_node_roles == ["open"]:
-        open_nodes.append(old_node)
-        continue
-      for _, node in enumerate(nodes_copy):
-        # Match nodes based on jobs/roles and the instance type specified.
-        if set(old_node_roles) == set(node.roles) \
-                and old_node.get('instance_type') == node.instance_type:
-          nodes_copy.remove(node)
-          node.from_json(old_node)
-          if node.is_valid():
-            nodes.append(node)
-          else:
-            # Locations JSON is incorrect if we get here.
-            return None
-          break
-    for open_node in open_nodes:
-      for node in nodes_copy:
-        # Match nodes based on jobs/roles and the instance type specified.
-        if node.instance_type == open_node.get('instance_type'):
-          roles = node.roles
-          node.from_json(open_node)
-          node.roles = roles
-          if node.is_valid():
-            nodes.append(node)
-          else:
-            # Locations JSON is incorrect if we get here.
-            return None
-        else:
-          continue
+    assert len(locations_nodes_list) == len(self.nodes), (
```

Comment:
Since Python can easily be configured to ignore assert statements, it's usually better to raise an exception. (my fault for suggesting the assert)